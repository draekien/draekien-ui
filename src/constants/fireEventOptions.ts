export const options: {} = {
  bubbles: true,
  cancelable: true,
};
