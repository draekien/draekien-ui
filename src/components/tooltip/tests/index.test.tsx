/** @jsx jsx */
import { jsx } from 'theme-ui';
import { act, fireEvent, render } from '@testing-library/react';
import { options } from '../../../constants/fireEventOptions';
import Tooltip from '..';

const Element = () => (
  <div id="test" style={{ width: '100px', height: '100px' }} />
);

describe('Tooltip component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  test('renders correctly when not hovered', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('div').length).toEqual(2);
  });

  test('should show tooltip on mouseEnter', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.mouseEnter(container.querySelectorAll('div')[0], options);

    act(() => {
      jest.runAllTimers();
    });

    expect(container.querySelectorAll('div').length).toEqual(4);
  });

  test('should show tooltip on focus', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.focus(container.querySelectorAll('#test')[0], options);

    act(() => {
      jest.runAllTimers();
    });

    expect(container.querySelectorAll('div').length).toEqual(4);
  });

  test('should wait 400ms before showing tooltip', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.focus(container.querySelectorAll('div')[0], options);

    expect(container.querySelectorAll('div').length).toEqual(2);

    act(() => {
      jest.runTimersToTime(400);
    });

    expect(container.querySelectorAll('div').length).toEqual(4);
  });

  test('should hide tooltip when mouseLeave', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.mouseEnter(container.querySelectorAll('div')[0], options);

    act(() => {
      jest.runAllTimers();
    });

    fireEvent.mouseLeave(container.querySelectorAll('div')[0], options);

    expect(container.querySelectorAll('div').length).toEqual(2);
  });

  test('should hide tooktip when blur', () => {
    const { container } = render(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.focus(container.querySelectorAll('div')[0], options);

    act(() => {
      jest.runAllTimers();
    });

    fireEvent.blur(container.querySelectorAll('div')[0], options);

    expect(container.querySelectorAll('div').length).toEqual(2);
  });

  test('should call callback functions for onBlur, onFocus, onMouseEnter and onMouseLeave', () => {
    const onBlur = jest.fn();
    const onFocus = jest.fn();
    const onMouseEnter = jest.fn();
    const onMouseLeave = jest.fn();

    const { container } = render(
      <Tooltip
        text="tooltip"
        onBlur={onBlur}
        onFocus={onFocus}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        <Element />
      </Tooltip>
    );

    const element = container.querySelectorAll('div')[0];

    fireEvent.focus(element, options);

    act(() => {
      jest.runAllTimers();
    });

    fireEvent.blur(element, options);

    act(() => {
      jest.runAllTimers();
    });

    fireEvent.mouseEnter(element, options);

    act(() => {
      jest.runAllTimers();
    });

    fireEvent.mouseLeave(element, options);

    act(() => {
      jest.runAllTimers();
    });

    expect(onBlur).toBeCalledTimes(1);
    expect(onFocus).toBeCalledTimes(1);
    expect(onMouseEnter).toBeCalledTimes(1);
    expect(onMouseLeave).toBeCalledTimes(1);
  });

  test('should display correct text in tooltip', () => {
    const { container } = render(
      <Tooltip text="this is tooltip">
        <Element />
      </Tooltip>
    );

    fireEvent.focus(container.querySelectorAll('div')[0], options);

    act(() => {
      jest.runAllTimers();
    });

    expect(container.textContent).toContain('this is tooltip');
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
});
