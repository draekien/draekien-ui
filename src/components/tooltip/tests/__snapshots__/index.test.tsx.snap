// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component renders correctly when not hovered 1`] = `
<div>
  <div
    class="css-h0bkxn-Tooltip"
    tabindex="1"
  >
    <div
      id="test"
      style="width: 100px; height: 100px;"
    />
  </div>
</div>
`;
